src/runtime_src/core/pcie/driver/linux/xocl/subdev/xdma.c

```diff
@@ -92,6 +92,110 @@ static ssize_t xdma_migrate_bo(struct platform_device *pdev,
 	return ret;
 }
 
+struct xdma_async_context {
+	void (*callback_fn)(unsigned long data, int err);
+	unsigned long callback_data;
+	struct xdma_io_cb *iocb;
+	struct xocl_xdma *xdma;
+	u32 dir;
+	u32 channel;
+};
+
+static void xdma_async_migrate_done(unsigned long data, int err)
+{
+	struct xdma_async_context *async_ctx = (struct xdma_async_context *)data;
+	//pr_info("%s: async_ctx %llx, ", __func__, (u64)async_ctx);
+
+	if (!err) {
+		async_ctx->xdma->channel_usage
+			[async_ctx->dir][async_ctx->channel] += 
+						async_ctx->iocb->done_bytes;
+	}
+	
+	async_ctx->callback_fn(async_ctx->callback_data, err);
```

Comment:
It will be good to have a check for callback func like if (async_ctx->callback_fn). Sometimes clients for this async method might not want completions notifications for each transfers.